### ISSUE TITLE
Type inference not working on Array.prototype.reduce

### ISSUE THREAD

**Issue by carneiror** [microsoft/TypeScript#33086]
**TypeScript Version:**  3.5.1 (tested on 3.6.1-rc also)

**Search Terms:**
- reduce generic
- spread operator

**Code**

```ts
interface Animal {
    name: string,
    id: string
}

interface State {
    [key: string]: Animal
}

const store: State = {};

function reducer(state: State, animals: Animal[]): State {
    return animals.reduce<State>((acc, animal) => ({
        [animal.id]: {
            ...state[animal.id],
            ...animal,
            random: 39846 // <-- Should throw an error since random is not a valid prop
        }
    }), { ...state } 
    );
}

reducer(store, [{
    id: "random-id",
    name: "Elephant"
    random: 387456 // <-- Correctly identify the invalid prop
}]);
```

**Expected behavior:** 

Types should be inferred on both examples 

**Actual behavior:**

Using reduce generic should enforce that the return should be of type `State` and identify `random` as an invalid prop of `Animal`.

**Playground Link:**

[Typescript Playground Link](http://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgIImAWzgG2QbwChkTkQ5MIAuZAZzClAHMAaY04AExvsZCcIBfQoVCRYiFAGUwcSAXYkA2gGsIATx4NmAXRrosuISIQB7EPTphTUashlyUAXgKCA3CJgBXEAjDBzZFtOLyQoAAp6RxoHSBZkOAxsHFp9JNwlHQBKGNl5IlIgiDAvKBAE9JSAOmDQiAAeWIgAPnDwxAR4xMMcLOQnZuRwgsLSJW7kqq49BVG55CrFqMhxyqnOHTZ50cWqidwt7dIoRM5TTBoAZgBOAA4AFgA2ZAB6F+R6gFpP+wALUy8OE4yDAvygpgA7hVkNBwVA6KAkEFTudkMBaGRTGAEsgAG64LjIAAO4KJilGwkKgiy8XwCyWeRQgmQ5KyHmEhFqYUi1ls8SUIxIXBoACITiAzphPlwRYcSORKKKAKI4CBE36JMAi8niyVXW4AdnuAFZnm8Pt9kABhGy2Pw4dRozgQcDAGCO0EoUD4nCEkmmMmCbJuIA)
